### PR TITLE
[mask_rom] Add Address Translation.

### PR DESCRIPTION
**Pull request by engdoreis**
# Introdution
Currently we generate two binaries for ROM_EXT, one statically linked for slot A and one statically linked for slot B. Ideally we would like to generate a single ROM_EXT binary that will work regardless of which position in flash it is placed in.

This PR intruduces the Ibex Simple Address Translation feature in the `MASK_ROM` to allow the `ROM_EXT` be pisition independent in case it is signalized in the manifest.

See issue https://github.com/lowRISC/opentitan/issues/9930
See the [proposal](https://docs.google.com/document/d/1wscqObX895t9Nbd2mPZCC7gUJL_wJLUaaMgBAT_sYhg/edit#heading=h.yqwfe0dy7w4k) for more detail.

### CODE REVIEW

File: sw/device/silicon_creator/lib/drivers/address_translation.c
Comment (mundaym) [inline sw/device/silicon_creator/lib/drivers/address_translation.c:20]: I think I'd be tempted to do this slot -> addr mapping in a switch statement in the function rather than using structs. Switch statements are easier to harden later if we need to.

File: sw/device/silicon_creator/lib/drivers/address_translation.h
Comment (mundaym) [inline sw/device/silicon_creator/lib/drivers/address_translation.h:43]: I can't think of a situation where we'd want to configure different I and D regions in silicon creator firmware. Maybe we can just get away with a single address translation pair.

**Review by mundaym — COMMENTED** [review]
Thanks @engdoreis. I think this is along the right lines. I've left a couple of minor comments.

Please proceed with integrating this into the mask ROM code.

File: sw/device/silicon_creator/lib/drivers/address_translation.h
Comment (engdoreis) [reply sw/device/silicon_creator/lib/drivers/address_translation.h:43]: Done

File: sw/device/silicon_creator/lib/drivers/address_translation.c
Comment (engdoreis) [reply sw/device/silicon_creator/lib/drivers/address_translation.c:20]: Done

**Comment by engdoreis** [comment]
Thanks @mundaym for the review. I have updated the driver and integrated it into the mask_rom. It is not ready yet. But if you could quickly pass I would appreciate it.

File: sw/device/silicon_creator/lib/manifest.h
Comment (mundaym) [inline sw/device/silicon_creator/lib/manifest.h:355]: I don't think we need this since we only need to know the slot we are unlocking for address translation configuration. Is that correct?

Also the region we unlock needs to be NAPOT aligned, the code section we use for ePMP may well not be.

File: sw/device/silicon_creator/mask_rom/mask_rom.c
Comment (mundaym) [inline sw/device/silicon_creator/mask_rom/mask_rom.c:217]: Related to my comment above I think it would be good to pass around a hardened enum specifying which slot the manifest/rom_ext is in rather than trying to extract address information from the manifest.

File: sw/device/silicon_creator/mask_rom/mask_rom.ld
Comment (mundaym) [inline sw/device/silicon_creator/mask_rom/mask_rom.ld:29]: We should already have this information since we are able to locate the manifests for both slots.

Comment (engdoreis) [reply sw/device/silicon_creator/mask_rom/mask_rom.ld:29]: The name is representing right, this address is the Virtual address (VMA). The address we have and matches with the manifest is the physical address (LMA).

File: sw/device/silicon_creator/lib/manifest.h
Comment (mundaym) [reply sw/device/silicon_creator/lib/manifest.h:355]: Sorry, I misunderstood what this was for.

File: sw/device/silicon_creator/lib/drivers/address_translation.c
Comment (mundaym) [inline sw/device/silicon_creator/lib/drivers/address_translation.c:15]: One thing that occurs to me is that we will only be configuring slot 0 in the mask ROM and so the slot 1 code path will never be used. Maybe for this reason we should have a function per slot rather than using an enum?

File: sw/device/silicon_creator/mask_rom/mask_rom.c
Comment (mundaym) [inline sw/device/silicon_creator/mask_rom/mask_rom.c:224]: I think we should make this an inline function rather than a macro but otherwise lgtm.

File: sw/device/silicon_creator/lib/drivers/address_translation.c
Comment (engdoreis) [reply sw/device/silicon_creator/lib/drivers/address_translation.c:15]: done

File: sw/device/silicon_creator/rom_ext/rom_ext_slot_b.ld
Comment (engdoreis) [inline sw/device/silicon_creator/rom_ext/rom_ext_slot_b.ld:13]: I'm going to remove this file in a future PR.

**Comment by alphan** [comment]
@moidx @cfrantz

**Comment by tjaychen** [comment]
just to chime in this area of the hardware has not been thoroughly verified (since it falls into the crack between ibex and rv_core_ibex), so let us know if you see anything suspicious.

File: sw/device/silicon_creator/mask_rom/mask_rom.c
Comment (alphan) [inline sw/device/silicon_creator/mask_rom/mask_rom.c:201]: Is there a reason why we want to unlock the whole rom_ext area instead of the area used by the rom_ext image (as given by `manifest->length`)?

**Review by alphan — COMMENTED** [review]
Thanks @engdoreis! This looks pretty good. I think we need to move some code around and think a little bit more about the linker scripts.

File: sw/device/silicon_creator/lib/manifest.h
Comment (engdoreis) [reply sw/device/silicon_creator/lib/manifest.h:235]: done

File: sw/device/silicon_creator/lib/drivers/address_translation.h
Comment (engdoreis) [reply sw/device/silicon_creator/lib/drivers/address_translation.h:35]: Done

File: sw/device/silicon_creator/lib/drivers/address_translation.h
Comment (engdoreis) [reply sw/device/silicon_creator/lib/drivers/address_translation.h:37]: Done

File: sw/device/silicon_creator/lib/drivers/address_translation_unittest.cc
Comment (engdoreis) [reply sw/device/silicon_creator/lib/drivers/address_translation_unittest.cc:27]: Done

File: sw/device/silicon_creator/lib/drivers/meson.build
Comment (engdoreis) [reply sw/device/silicon_creator/lib/drivers/meson.build:492]: Done

File: sw/device/silicon_creator/lib/drivers/meson.build
Comment (engdoreis) [reply sw/device/silicon_creator/lib/drivers/meson.build:525]: done

File: sw/device/silicon_creator/mask_rom/docs/memory_protection.md
Comment (engdoreis) [reply sw/device/silicon_creator/mask_rom/docs/memory_protection.md:92]: done

File: sw/device/silicon_creator/mask_rom/mask_rom.c
Comment (engdoreis) [reply sw/device/silicon_creator/mask_rom/mask_rom.c:209]: Done

File: sw/device/silicon_creator/mask_rom/mask_rom.c
Comment (engdoreis) [reply sw/device/silicon_creator/mask_rom/mask_rom.c:248]: Done

File: sw/device/silicon_creator/mask_rom/mask_rom.c
Comment (engdoreis) [reply sw/device/silicon_creator/mask_rom/mask_rom.c:250]: done

File: sw/device/silicon_creator/mask_rom/mask_rom.c
Comment (engdoreis) [reply sw/device/silicon_creator/mask_rom/mask_rom.c:201]: I think that the main reason is that the size must be a power of two (NAPOT), then there's a great change of the `manifest->length` ending up being rounded to the max size anyway.

File: sw/device/silicon_creator/mask_rom/mask_rom.c
Comment (engdoreis) [reply sw/device/silicon_creator/mask_rom/mask_rom.c]: I think that the main reason is that the size must be a power of two (NAPOT), then there's a great change of the `manifest->length` ending up being rounded to the `_rom_ext_virtual_size` anyway.

File: sw/device/silicon_creator/lib/drivers/address_translation.h
Comment (engdoreis) [reply sw/device/silicon_creator/lib/drivers/address_translation.h]: I don't have a strong opinion on that, I think both works well, althought `size_t` looks more "semanticaly right".

File: sw/device/silicon_creator/rom_ext/rom_ext_common.ld
Comment (engdoreis) [reply sw/device/silicon_creator/rom_ext/rom_ext_common.ld]: I'm keeping the build of the two slots, and I added a new target linked to the virtual address.

File: sw/device/silicon_creator/rom_ext/rom_ext_slot_a.ld
Comment (alphan) [reply sw/device/silicon_creator/rom_ext/rom_ext_slot_a.ld]: Resolving this, discussed offline.

File: hw/top_earlgrey/sw/autogen/top_earlgrey_memory.ld
Comment (alphan) [reply hw/top_earlgrey/sw/autogen/top_earlgrey_memory.ld:16]: Captured in #12596

File: sw/device/silicon_creator/mask_rom/mask_rom_epmp.h
Comment (alphan) [inline sw/device/silicon_creator/mask_rom/mask_rom_epmp.h:69]: nit: this sentence doesn't look right

File: sw/device/silicon_creator/mask_rom/mask_rom_epmp.h
Comment (alphan) [inline sw/device/silicon_creator/mask_rom/mask_rom_epmp.h:63]: nit: I think the comment should mention that this entry uses NAPOT.

File: sw/device/silicon_creator/rom_ext/BUILD
Comment (alphan) [inline sw/device/silicon_creator/rom_ext/BUILD:156]: nit: we should probably call this "addr_translation" or "virtual_addr" so that it doesn't conflict with the actual PIC when we have it.

File: sw/device/silicon_creator/rom_ext/BUILD
Comment (alphan) [inline sw/device/silicon_creator/rom_ext/BUILD:97]: nit: "va" suffix is a bit cryptic. WDYT about `rom_ext_virtual_addr.ld`?

File: sw/device/silicon_creator/mask_rom/mask_rom.ld
Comment (engdoreis) [reply sw/device/silicon_creator/mask_rom/mask_rom.ld:32]: done

File: sw/device/silicon_creator/mask_rom/mask_rom_epmp.h
Comment (engdoreis) [reply sw/device/silicon_creator/mask_rom/mask_rom_epmp.h:63]: done

File: sw/device/silicon_creator/rom_ext/BUILD
Comment (engdoreis) [reply sw/device/silicon_creator/rom_ext/BUILD:97]: done

File: sw/device/silicon_creator/mask_rom/mask_rom_epmp.h
Comment (engdoreis) [reply sw/device/silicon_creator/mask_rom/mask_rom_epmp.h:63]: line 67

File: sw/device/silicon_creator/lib/drivers/BUILD
Comment (alphan) [inline sw/device/silicon_creator/lib/drivers/BUILD:188]: nit: seems like a tab/space issue

File: sw/device/silicon_creator/mask_rom/mask_rom.c
Comment (alphan) [reply sw/device/silicon_creator/mask_rom/mask_rom.c:201]: Makes sense, I forgot that entire eFlash is unlocked for read anyway.

File: sw/device/silicon_creator/mask_rom/mask_rom.c
Comment (alphan) [inline sw/device/silicon_creator/mask_rom/mask_rom.c:263]: nit: addres -> address

File: sw/device/silicon_creator/mask_rom/mask_rom_epmp.c
Comment (alphan) [inline sw/device/silicon_creator/mask_rom/mask_rom_epmp.c:105]: minor nit: Would 
```suggestion
void mask_rom_epmp_config_rom_ext_r(epmp_state_t *state, epmp_region_t region) {
```
or something similar be clearer since the area we unlock is greater than the image.

File: sw/device/silicon_creator/rom_ext/meson.build
Comment (alphan) [inline sw/device/silicon_creator/rom_ext/meson.build:54]: I think the file name needs to be updated.

File: sw/device/silicon_creator/rom_ext/rom_ext_virtual_addr.ld
Comment (alphan) [inline sw/device/silicon_creator/rom_ext/rom_ext_virtual_addr.ld:12]: This comment needs to be updated since this script is for address translation/virtual address.

File: sw/device/silicon_creator/rom_ext/rom_ext_virtual_addr.ld
Comment (alphan) [inline sw/device/silicon_creator/rom_ext/rom_ext_virtual_addr.ld:26]:
```suggestion
_rom_ext_start_address = ORIGIN(rom_ext_virtual);
_rom_ext_size = LENGTH(rom_ext_virtual);
```

**Review by alphan — APPROVED** [review]
Thanks @engdoreis!

Can you fix the CI issues and make sure that ROM can jump to ROM_EXT with and without address translation before merging this? Thanks!

**Comment by alphan** [comment]
I think we'll need to think a little bit about how we will use address translation for the first owner boot stage but that discussion orthogonal to this PR.

File: sw/device/silicon_creator/mask_rom/mask_rom.c
Comment (moidx) [inline sw/device/silicon_creator/mask_rom/mask_rom.c:245]: nit: The spacing of this comment is not aligned with the code block. May be an issue with whitespace.

File: sw/device/silicon_creator/mask_rom/mask_rom.c
Comment (moidx) [inline sw/device/silicon_creator/mask_rom/mask_rom.c:284]: Given the new set of register writes, it is probably a good idea to consider moving the `sec_mmio_check_values(rnd_uint32())` check right before this call.

**Review by moidx — APPROVED** [review]

File: sw/device/silicon_creator/lib/drivers/BUILD
Comment (engdoreis) [reply sw/device/silicon_creator/lib/drivers/BUILD:188]: Fixed

File: sw/device/silicon_creator/mask_rom/mask_rom.c
Comment (engdoreis) [reply sw/device/silicon_creator/mask_rom/mask_rom.c:263]: fixed

File: sw/device/silicon_creator/mask_rom/mask_rom_epmp.c
Comment (engdoreis) [reply sw/device/silicon_creator/mask_rom/mask_rom_epmp.c:105]: done

File: sw/device/silicon_creator/rom_ext/meson.build
Comment (engdoreis) [reply sw/device/silicon_creator/rom_ext/meson.build:54]: fixed

File: sw/device/silicon_creator/mask_rom/mask_rom.c
Comment (engdoreis) [reply sw/device/silicon_creator/mask_rom/mask_rom.c:245]: fixed

File: sw/device/silicon_creator/mask_rom/mask_rom.c
Comment (engdoreis) [reply sw/device/silicon_creator/mask_rom/mask_rom.c:284]: done

**Comment by engdoreis** [comment]
Thanks @alphan, @moidx and @mundaym for the reviews. 
I tested with and without the address translation enabled and it booted the rom_ext successfuly.
So if you guys are happy I think you can merge this PR once the CI finishes.

Should I open an Issue to disable the address translate on the manifest for the binaries rom_ext_slot_a and rom_ext_slot_b?

**Comment by alphan** [comment]
> Should I open an Issue to disable the address translate on the manifest for the binaries rom_ext_slot_a and rom_ext_slot_b?

Please do (and assign to @jon-flatley) but I don't think we can really address it at the moment since we are still using the legacy signer and it doesn't expose it on the CLI. If there are no failures, we can keep the default value for address_translation as true (and use the new rom_ext image that you added for any ad-hoc tests) until we start using opentitantool.

File: util/topgen/templates/toplevel_memory.ld.tpl
Comment (alphan) [reply util/topgen/templates/toplevel_memory.ld.tpl:54]: #12670

**Pull request merged**
